Add for Ruby3.x support

maedana opened the pull request.

mziyut reviewed (commented):
Thank you for the pull request!

Can you please fix the version problem of "activesupport" and "rubocop" in the file "qiita-markdown.gemspec" ?

Also, the Ruby v3.x compatibility guarantee for "greenmat" listed in the dependencies has been done in https://github.com/increments/greenmat/pull/18 . 
The upgrade of "greenmat" from 3.5.1.3 to 3.5.1.4 will be done by us.

mziyut commented:
On my Ruby v3.1.1 environment, "rubocop" fails when I run `bundle exec rake` (spec is succeeds).
It reason for the failure is that the "rubocop" version listed in the development dependencies of `qiita-markdown.gemspec` is unsupported version.

```
warning: parser/current is loading parser/ruby31, which recognizes3.1.2-compliant syntax, but you are running 3.1.1.
Please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
Running RuboCop...
.rubocop.yml: Style/PredicateName has the wrong namespace - should be Naming
.rubocop.yml: Metrics/LineLength has the wrong namespace - should be Layout
Error: RuboCop found unknown Ruby version 3.1 in `.ruby-version`.
Supported versions: 2.4, 2.5, 2.6, 2.7, 3.0
RuboCop failed!
```

Is it possible to update the version of "rubocop" (e.g. v1.27.0) as well as "activesupport"?

"rubocop" update will cause a new warning, please temporarily add it to your `.rubocop_todo.yml`. We will fix it as a separate task.

maedana commented:
@mziyut Thank you for your reply. I've fixed the two points you pointed out. I would be happy if you could check them.

> Is it possible to update the version of "rubocop" (e.g. v1.27.0) as well as "activesupport"?

I've updated `rubocop` and `.rubocop_todo.yml`, but some warnings remain.
```
❯ bundle exec rubocop                                      
.rubocop.yml: Style/PredicateName has the wrong namespace - should be Naming            
.rubocop.yml: Metrics/LineLength has the wrong namespace - should be Layout
Inspecting 47 files                                                                                                    
..............C......C....C....................            
                                                                                                                                                                                                                                              
Offenses:                                                                                                              
                                                           
lib/qiita/markdown/filters/code_block.rb:28:27: C: [Correctable] Style/HashSyntax: Omit the hash value.
                filename: filename,                   
                          ^^^^^^^^                                                                                     
lib/qiita/markdown/filters/code_block.rb:29:27: C: [Correctable] Style/HashSyntax: Omit the hash value.
                language: language,                                                                                    
                          ^^^^^^^^                                                                                     
lib/qiita/markdown/filters/group_mention.rb:65:41: C: [Correctable] Style/HashSyntax: Omit the hash value.
              group = { group_url_name: group_url_name, team_url_name: team_url_name }                                 
                                        ^^^^^^^^^^^^^^     
lib/qiita/markdown/filters/group_mention.rb:65:72: C: [Correctable] Style/HashSyntax: Omit the hash value.                                                                                                                                    
              group = { group_url_name: group_url_name, team_url_name: team_url_name }
                                                                       ^^^^^^^^^^^^^    
lib/qiita/markdown/filters/syntax_highlight.rb:15:35: C: [Correctable] Style/HashSyntax: Omit the hash value.          
                default_language: default_language,                                                                    
                                  ^^^^^^^^^^^^^^^^                                                                     
lib/qiita/markdown/filters/syntax_highlight.rb:16:23: C: [Correctable] Style/HashSyntax: Omit the hash value.                                                                                                                                 
                node: node,                          
                      ^^^^                                 
                                                                                                                       
47 files inspected, 6 offenses detected, 6 offenses auto-correctable                                                   
                                                           
Tip: Based on detected gems, the following RuboCop extension libraries might be helpful:
  * rubocop-rake (https://rubygems.org/gems/rubocop-rake)
  * rubocop-rspec (https://rubygems.org/gems/rubocop-rspec)                                                            
                                                           
You can opt out of this message by adding the following to your config (see https://docs.rubocop.org/rubocop/extensions.html#extension-suggestions for more options):
  AllCops:                                                                                                             
    SuggestExtensions: false                         
```

I think the old style looks fine. So I've added `EnforcedShorthandSyntax: never` to `Style/HashSyntax:`.
See also https://docs.rubocop.org/rubocop/cops_style.html#stylehashsyntax

```diff
 # SupportedShorthandSyntax: always, never, either
 Style/HashSyntax:
   EnforcedStyle: ruby19
+  EnforcedShorthandSyntax: never
```

mziyut commented:
@maedana Thanks for the correction!

mziyut commented on .rubocop_todo.yml:
```diff
@@ -181,15 +220,12 @@ Style/MutableConstant:
     - 'lib/qiita/markdown/filters/code_block.rb'
-    - 'lib/qiita/markdown/filters/group_mention.rb'
-    - 'lib/qiita/markdown/filters/inline_code_color.rb'
-    - 'lib/qiita/markdown/filters/mention.rb'
     - 'lib/qiita/markdown/filters/simplify.rb'
```
I verified the changes with the development environment and CI (GitHub Actions).
I would like to ask you to correct again some differences in the generated `.rubocop_todo.yml`.

```diff
diff --git a/.rubocop_todo.yml b/.rubocop_todo.yml
index 84d407e..3263685 100644
--- a/.rubocop_todo.yml
+++ b/.rubocop_todo.yml
@@ -210,7 +210,7 @@ Style/IfUnlessModifier:
     - 'lib/qiita/markdown/filters/truncate.rb'
     - 'lib/qiita/markdown/transformers/strip_invalid_node.rb'

-# Offense count: 14
+# Offense count: 17
 # This cop supports safe auto-correction (--auto-correct).
 # Configuration parameters: EnforcedStyle.
 # SupportedStyles: literals, strict
@@ -220,6 +220,9 @@ Style/MutableConstant:
     - 'lib/qiita/markdown/embed/tweet.rb'
     - 'lib/qiita/markdown/filters/checkbox.rb'
     - 'lib/qiita/markdown/filters/code_block.rb'
+    - 'lib/qiita/markdown/filters/group_mention.rb'
+    - 'lib/qiita/markdown/filters/inline_code_color.rb'
+    - 'lib/qiita/markdown/filters/mention.rb'
     - 'lib/qiita/markdown/filters/simplify.rb'
     - 'lib/qiita/markdown/filters/syntax_highlight.rb'
     - 'lib/qiita/markdown/version.rb'
```

mziyut replied:
Thanks for the fix! 👍

mziyut reviewed (commented):
Thanks for the correction!
Sorry for the inconvenience, could you please make one more fixed?

mziyut commented:
CI is failing because Token cannot be referenced during Fork.
We will fix it separately so that CI can be executed when you fork the project.

mziyut reviewed (approved):
Thank you! ✅ GTM!

mziyut commented:
Released in https://github.com/increments/qiita-markdown/pull/119 .
Thank you!!

Ref: [qiita-markdown | RubyGems.org | your community gem host](https://rubygems.org/gems/qiita-markdown/versions/0.42.0)
